examples/gpact/hotel-train/contracts/src/Train.sol

```diff
@@ -0,0 +1,201 @@
+    // the spending address can't be changed. If this functionality
+    // was required, the system would need to be carefully designed
+    // for locking.
+    // Map (address of travel agency contract => spender address for the travel agency)
+    mapping(address => address) travelAgencySpender;
+    // Map (blockchain id => address of travel agency contract)
+    mapping(uint256 => address) approvedTravelAgencies;
+
+    address owner;
+    IERC20 erc20;
+
+    modifier onlyOwner() {
+        require(msg.sender == owner);
+        _;
+    }
+
+    constructor(address _erc20, address _cbc) LockableStorage(_cbc) {
+        owner = msg.sender;
+        erc20 = IERC20(_erc20);
+    }
+
+    function addApprovedTravelAgency(
+        uint256 _blockchainId,
+        address _travelAgencyContract,
+        address spendingAccount
+    ) external onlyOwner {
+        approvedTravelAgencies[_blockchainId] = _travelAgencyContract;
+        travelAgencySpender[_travelAgencyContract] = spendingAccount;
+    }
+
+    function addSeats(uint256 _seatRate, uint256 _numberOfSeats)
+        external
+        onlyOwner
+    {
+        for (uint256 i = 0; i < _numberOfSeats; i++) {
+            setMapValue(SEAT_RATE_MAP, numSeats++, _seatRate);
+        }
+    }
+
+    function changeSeatRate(uint256 _seatNumber, uint256 _seatRate)
+        external
+        onlyOwner
+    {
+        setMapValue(SEAT_RATE_MAP, _seatNumber, _seatRate);
+    }
+
+    function bookSeat(
+        uint256 _date,
+        uint256 _uniqueId,
+        uint256 _maxAmountToPay
+    ) external {
+        require(address(cbc) == msg.sender, "Must be crosschain call");
+
+        // Check that the calling contract was the travel agency linked to this one from
+        // the source blockchain.
+        // TODO check that the root blockchain is trusted
+        (
+            ,
+            uint256 sourceBlockchainId,
+            address sourceContract
+        ) = decodeAtomicAuthParams();
+        require(
+            sourceContract == approvedTravelAgencies[sourceBlockchainId],
+            "Sender is not an approved travel agency"
+        );
+
+        require(_date != 0, "Date can not be zero");
+        for (uint256 i = 0; i < numSeats; i++) {
+            uint256 rate = getMapValue(SEAT_RATE_MAP, i);
+            // If amount is OK and the room is available.
+            if (
+                !isDoubleMapValueLocked(SEAT_BOOKED_BY_2MAP, i, _date) &&
+                rate <= _maxAmountToPay &&
+                address(
+                    uint160(getDoubleMapValue(SEAT_BOOKED_BY_2MAP, i, _date))
+                ) ==
+                address(0)
+            ) {
+                // Book seat
+                setDoubleMapValue(
+                    SEAT_BOOKED_BY_2MAP,
+                    i,
+                    _date,
+                    uint160(tx.origin)
+                );
+                setMapValue(BOOKING_REF_TO_SEAT_NUMBER, _uniqueId, i);
```

Comment:
nit: this probably doesn't matter much given this is just example code, but should the uniqueness `_uniqueId` be checked here? can one overwrite another booking otherwise